File: 3612. Process String with Special Operations I.java
class Solution {

    public String processStr(String inputString) {

        // Stores the processed string
        StringBuilder processedString =
                new StringBuilder();

        for (char currentCharacter :
                inputString.toCharArray()) {

            // If it is a lowercase letter,
            // append it to the result
            if (currentCharacter >= 'a' &&
                currentCharacter <= 'z') {

                processedString.append(
                        currentCharacter
                );
            }

            // Remove the last character
            else if (currentCharacter == '*') {

                if (processedString.length() > 0) {

                    processedString.deleteCharAt(
                            processedString.length() - 1
                    );
                }
            }

            // Duplicate the current string
            else if (currentCharacter == '#') {

                processedString.append(
                        processedString.toString()
                );
            }

            // Reverse the current string
            else if (currentCharacter == '%') {

                processedString.reverse();
            }
        }

        return processedString.toString();
    }
}

File: 3612. Process String with Special Operations I.py
class Solution:
    def processStr(
        self,
        input_string: str
    ) -> str:

        # Using a list for efficient modifications
        processed_characters = []

        for current_character in input_string:

            # If it is a lowercase letter,
            # append it to the result
            if 'a' <= current_character <= 'z':

                processed_characters.append(
                    current_character
                )

            # Remove the last character
            elif current_character == '*':

                if processed_characters:

                    processed_characters.pop()

            # Duplicate the current string
            elif current_character == '#':

                processed_characters.extend(
                    processed_characters
                )

            # Reverse the current string
            elif current_character == '%':

                processed_characters.reverse()

        return "".join(processed_characters)
